### PR TITLE
libnetwork: make resolvconf more self-contained

### PR DESCRIPTION
**Pull request by thaJeztah**
### libnetwork: move resolvconf consts into the resolvconf package

This allows using the package without having to import the "types" package, and without having to consume github.com/ishidawataru/sctp.

### libnetwork: remove resolvconf/dns package

The IsLocalhost utility was not used, and go's "net" package provides a `IsLoopBack()` check, which can be used in stead of `IsIPv4Localhost`.

### move pkg/ioutils.HashData() to libnetwork/resolvconf

It's the only location it's used, so we might as well move it there. This also removes the "crypto/sha256" and "encoding/hex" dependencies from pkg/ioutils..

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
relates to https://github.com/moby/buildkit/pull/2317

**Comment by thaJeztah** [comment]
Something broke, but I'm confused what. I suspect it's related to the pkg/ioutils.HashData() utility, but that was just _moved_ 🤔 ;

```
=== Failed
=== FAIL: reference TestLoad (0.00s)
    store_test.go:53: failed to parse reference: unsupported digest algorithm

=== FAIL: reference TestSave (0.00s)
    store_test.go:82: failed to parse reference: unsupported digest algorithm

=== FAIL: reference TestAddDeleteGet (0.00s)
    store_test.go:174: could not parse reference: unsupported digest algorithm

=== FAIL: reference TestInvalidTags (0.00s)
    store_test.go:355: assertion failed: error is not nil: unsupported digest algorithm
```

**Comment by thaJeztah** [comment]
Opened https://github.com/opencontainers/go-digest/pull/64 to fix go-digest

File: pkg/ioutils/readers.go [hunk 1]
```diff
@@ -2,9 +2,12 @@ package ioutils // import "github.com/docker/docker/pkg/ioutils"
...
 import (
 	"context"
-	"crypto/sha256"
-	"encoding/hex"
 	"io"
+
+	// make sure crypto.SHA256, crypto.sha512 and crypto.SHA384 are registered
+	// TODO remove once https://github.com/opencontainers/go-digest/pull/64 is merged.
+	_ "crypto/sha256"
+	_ "crypto/sha512"
```
Comment (thaJeztah) [inline pkg/ioutils/readers.go:10]: Included a temporary fix here (until https://github.com/opencontainers/go-digest/pull/64 is merged)

File: libnetwork/sandbox_dns_unix.go [hunk 2]
```diff
@@ -172,7 +172,9 @@ func (sb *sandbox) setExternalResolvers(content []byte, addrType int, checkLoopb
 	for _, ip := range servers {
 		hostLoopback := false
 		if checkLoopback {
-			hostLoopback = dns.IsIPv4Localhost(ip)
+			if p := net.ParseIP(ip); p != nil {
```
Comment (cpuguy83) [inline libnetwork/sandbox_dns_unix.go:175]: Isn't this different since it could be an ipv6 loopback?

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: libnetwork/sandbox_dns_unix.go (hunk 2 above)
Comment (thaJeztah) [reply libnetwork/sandbox_dns_unix.go:175]: Ah, yes. Brainfart; I was thinking this was to _skip_ loopback IP-addresses (as they won't work as DNS for container), but this is the bit where we forward requests to them. Yup, that shouldn't use IPv6 likely. Or maybe it should, but we can do that separately.

**Comment by thaJeztah** [comment]
@cpuguy83 updated; PTAL

**Review by cpuguy83 — APPROVED** [review]
😝 LGTM

File: libnetwork/sandbox_dns_unix.go [hunk 3]
```diff
@@ -181,6 +181,18 @@ func (sb *sandbox) setExternalResolvers(content []byte, addrType int, checkLoopb
 	}
 }
 
+// isIPv4Loopback checks if the given IP address is an IPv4 loopback address.
+// It's based on the logic in Go's net.IP.IsLoopback(), but only the IPv4 part:
+// https://github.com/golang/go/blob/go1.16.6/src/net/ip.go#L120-L126
+func isIPv4Loopback(ipAddress string) bool {
```
Comment (tianon) [inline libnetwork/sandbox_dns_unix.go:187]: This is a lot less eyebrow raising than the regular expressions above for parsing IP addresses in a language with a pretty decent standard library for parsing IP addresses. 😅 👍

**Review by tianon — APPROVED** [review]

File: libnetwork/sandbox_dns_unix.go (hunk 3 above)
Comment (thaJeztah) [reply libnetwork/sandbox_dns_unix.go:187]: Ha! True. Yes, I didn't compare performance between this and the original approach, but I assume the stdlib code is usually optimised.

**Comment by thaJeztah** [comment]
Failure is unrelated, and a known flaky test ; https://github.com/moby/moby/issues/36907

```
=== RUN   TestCreateServiceConfigFileMode
    create_test.go:355: assertion failed: 2 (int) != 1 (int)
--- FAIL: TestCreateServiceConfigFileMode (8.57s)
```

**Pull request merged**